WIP - Html introduction and basic exercise (#3)
FACN8/Pre-course: Week1/HTML.md

# Introduction to HTML

![html5 logo](https://user-images.githubusercontent.com/22002193/69446950-26f68b00-0d5e-11ea-96dd-eb217f407b56.jpg)

HTML is one of the most widely used languages on Web to develop web pages. It helps you delve into the world of Web Development and improve your skills.

## Contents

- Origin of HTML
- What is HTML?
  - Tag exploration
- HTML Structure
- How does HTML work?
- Fundamentals of HTML

## Origin of HTML

HTML - Hyper Text Markup Language is a language that was created by Tim Berners-Lee. Berners-Lee was putting together his first elementary browsing and authoring system for the Web and created a quick little hypertext language that would serve his purposes.

HTML was text-based and anyone could use any editor or word processor to create or convert documents for the Web. The developers started implementing new features in their browsers and started releasing advanced versions of HTML.

## What is HTML?

A markup language is a computer language that is used to apply layout and formatting conventions to a text document. Markup language makes the text more interactive and dynamic. It can turn text into images, tables, links, etc.

> ![html blocks](https://user-images.githubusercontent.com/22002193/69431669-dde40e00-0d40-11ea-9eca-eebebf65e58f.jpg)

Reviewer: is this image actually helpful for what is html ? cause even knowing html, not sure how it's related

1. The `<html></html>` tag is the highest level element that encloses every HTML page.
2. The `<head></head>` tag holds meta information such as the page’s title and charset.
3. Finally, the `<body></body>` tag encloses **all the content that appears on the page**.

## How does HTML work?

HTML documents end with the .html or .htm extension. You can view it using any web browser. The browser reads the HTML file and renders the content for users to view it.

### Task: Build a basic HTML page

- open your favourite editor and create a file called `my-first-page.html`
- type `html` in the page and press the `tab` key, this will automatically generate an html page for you.

The result should look like the following:

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title></title>
  </head>
  <body></body>
</html>
```

_Note: We will explore all the tags later_

- In the `<body>` tag add an `<h1>`

Headings are defined with the `<h1>` to `<h6>` tags.

`<h1>` defines the most important heading and `<h6>` defines the least important heading.

Your code should look something like this:

```html
<body>

>   <h1>The Main Heading</h1>

Reviewer: instead of showing a code, maybe we should show the output or the result of the code,

Author: They should see the result themselves, the point is for them to understand what html does and how you can write it

Reviewer: yeah but then what stops me from copy pasting? if you dont mind that then fine i guess

Author: It's okay if they copy-paste at this point

</body>
```

- Right click on your html file and click on `open in browser`, this should open the browser with the result of this file. You should see the h1 heading with `The Main Heading` on the page.

<details>

>   <summary>If you are using Atom</summary>

Reviewer 2: I think we (_Mario cough cough_) have encouraged the students to install VSCode, in the Ubuntu workshop. Shouldn't we add instructions for VSCode users here?

Reviewer 3: To the best of my understanding, Shireen's original instructions (right click) are the default in VSCode.

  To be able to preview your html file in a browser, install the _open-in-browsers_ package by going to:
  
  Atom -> Preferences -> Install
  </details>

- Let's explore this a bit more, in the browser, open the inspector. You can do so by either pressing `f12` or right clicking on the page and then inspect or pressing `ctrl+shift+i`.

>   - You should see be able to see the elements you created before (raw html), in the elements tab. You can edit/delete/add elements here like you did in your editor.

Reviewer 2: A screenshot of the elements tab here could be helpful?

Reviewer 3: Okay, will do.

>   - You should see be able to see the elements you created before (raw html), in the elements tab. You can edit/delete/add elements here like you did in your editor.

Reviewer: `You should see be able to see ` :sweat_smile:

Reviewer 3: ```suggestion
  - You should be able to see the elements you created before (raw html), in the elements tab. You can edit/delete/add elements here like you did in your editor.
```

>   - Change the elements in the browser from the inspector, delete the current `h1` tag and add a `div` tag.

Reviewer: why introduce this process ? i think showing how to add and delete from VSCode is enough, no need to confuse them with 2 editors.

Author: They should know that there are a few options and that the inspector is not just a way to see the elements but that you can also manipulate things from. 
It's not that confusing

Reviewer: it is very confusing since you almost never going to manually add an html element to the browser, and it is introducing a new development environment. why not just tell them to add the div in VSCode ?

Author: @kira-abl, after some debate we agreed to remove the editing from the browser.

- keep the note
- they should add a div from the editor and then continue to the other steps

Reviewer 3: @shiryz  So the only change will be that they will add a div in their editor instead of the browser? All the rest stays the same?

Author: Yes

Reviewer 3: @shiryz Done.

>     Find out more about [`div`](https://www.w3schools.com/tags/tag_div.asp)

Reviewer: here maybe we should do something like this 
> [click here](https://www.w3schools.com/tags/tag_div.asp) to find out more about  what is a div

Reviewer: but if you remove the inspector part where you add the div then maybe we should remove this part too

Author: Keep it consistent with the rest of the file, same as below

Reviewer: yeah it just seems out of place

Reviewer 3: @shiryz So do we keep it like this (here and everywhere) or do we change it?

Author: Following the comment above it should stay and would be similar to this:
> Read about HTML [links](https://www.w3schools.com/html/html_links.asp)

>     _Notice_ that these changes will not be saved in your file and only affect the browser, if you wish to save them then you will have to actually do it from the file itself.

Reviewer: and this comment with it too

- Add some more elements to in your `div`, after each step check your browser to see the changes:

>   - Add the `h1` tag back to your code with the content `The Main Heading`.

Reviewer: this is confusing since we never removed the h1 from the code, we removed it just from the browser. and i guess this task is confusing in general because im not sure where i need to do it, in the browser or in VSCode

>   - Add an `h2` tag that has the content `Sunheading`.  

Reviewer 2: Do you mean `Subheading` rather than `Sunheading`? 🌞

Reviewer 3: Yes, she did.

>     Why are headings [important?](https://www.w3schools.com/html/html_headings.asp)

Reviewer: >  Why are headings [important?]

Author: What was the point here?

Reviewer: same as this https://github.com/FACN8/Pre-course/pull/3#discussion_r350167606

  - Add a `p` tag with the content `First paragraph`.

>   - Add an `img` tag with an image from your laptop of your choosing.  

Reviewer: it's a new installation of Ubuntu there is probably going to be no local images :sweat_smile: ,
so maybe have a link instead ?

Reviewer 3: ```suggestion
  - Add an `img` tag with an image of your choosing (find an image online and use its link as src).  
```

>   - Add an `img` tag with an image from your laptop of your choosing.  

Reviewer 2: Have we explained at this point that an `<img>` tag doesn't need a closing tag, ie: it's self closing? Might be worth explaining that some tags don't require closing tags and / or can self close if there's no content inside them eg: `<br />`

Reviewer 3: @jamiecoe Thanks for the suggestion, implemented.

>     More about [`img`](https://www.w3schools.com/tags/tag_img.asp).

Reviewer: again in a comment
 > More about [`img`](https://www.w3schools.com/tags/tag_img.asp).

  - Add a second `div` (nested div).
  - In the second `div` add another `p` tag with the content `Second paragraph`.  
    
    Comments are used to leave messages for other devs without affecting the code itself. To add a comment in an html           file use the following syntax: `<!--` your comment goes in here `-->`
  - Now add a comment to your file just below the main heading.
  
  - The result should look something like this (_only use this in case you are stuck_):

    <details>

>      <summary>Code result</summary>

Reviewer: have this in ```html

so it looks like this

```html
<p>Hello this is an actual html </p>
```

Author: You can't add it in the details tag

Reviewer: i just tried you can wrap the everything in ````html  some html here> ```

       &lt;div&gt;
       <br>
        &lt;h1&gt;The Main Heading&lt;/h1&gt;
        <br>
        &lt;!-- My smart comment! --&gt;
        <br>
        &lt;h2&gt;Subheading&lt;/h2&gt;
        <br>
        &lt;p&gt;Paragraph&lt;/p&gt;
        <br>
        &lt;img src="/" alt="Image"&gt;
        <br>
        &lt;div&gt;
        <br>
          &lt;p&gt;Second Paragraph&lt;/p&gt;
          <br>
        &lt;/div&gt;
        <br>
      &lt;/div&gt;
    </details>

      <details>
      <summary>Result in browser</summary>
        <div>
          <h1>The Main Heading</h1>
          <h2>Subheading</h2>
          <p>Paragraph</p>

>           <img src="/" alt="Image">

Reviewer: the image here is not actually working :sweat_smile:

Author: They should add their own link

Reviewer: what ? this is supposed to be the result that they should see.

Author: Yes, but we don't the image's source, it should be a link unless we ask for a specific one

          <div>
            <p>Second Paragraph</p>
          </div>
        </div>
      </details>
      
### Task 2: Add more elements to your HTML page  

- Read about ways of creating [ordered](https://www.w3schools.com/tags/tag_ol.asp) and [unordered](https://www.w3schools.com/tags/tag_ul.asp) lists in html. In your file add two new `div`s and create one list in each of them. The lists should look like this:  

My to-do list:
1. (_insert your tasks here_) 
2.  
3.  

  Three people sitting next to me:  
  - (_insert their names_)
  -  
  -  

> ### Task 3 - Build a webpage to match the mockup  

Reviewer: what is a mockup ?

Reviewer 3: ```suggestion
### Task 3 - Build a webpage to match the _mockup_
( In design, a mockup is  a mid- to high-fidelity representation of the product’s appearance that shows the basics of its functionality)   
```

> Create a new HTML file (let's call it exercise.html) and try to create a webpage that will look exactly like this mockup (you can click on the image to enlarge it). Mind the instructions.

Reviewer: `(you can click on the image to enlarge it)` you can't :man_shrugging:

Reviewer 3: Well, _I_ can.

Reviewer 3: ```suggestion
Create a new HTML file (let's call it exercise.html) and try to create a webpage that will look exactly like this mockup (you can click on the image to open it in a new tab, if you'd like to see a larger version of it). Mind the instructions.
```

![](https://i.imgur.com/UsMwmXa.png)

##### Instructions  

> - Use only HTML, no CSS.

Reviewer: isn't it a given why mention no CSS?

Reviewer 3: It might not be a given for those who know what CSS is and might be inclined to use it.

Reviewer: i meant it's a given that they shouldn't use it. since they weren't told to and it's an html workshop

> - You can choose another picture of Narhwal, this picture is just an example.

Reviewer: ok this is another reason to do the comment the i mentioned up there
> it's a new installation of Ubuntu there is probably going to be no local images sweat_smile ,
so maybe have a link instead ?

- The picture should be a link leading to an article about Narwhal (use whatever article you would like).
- The name _Mac's Mission_ should lead to https://www.iampitbullstrong.com//.  

<details>
     <summary>Hints</summary>  
  
  - Read about HTML [links](https://www.w3schools.com/html/html_links.asp)  
  
  - Read about HTML [tags](https://www.w3schools.com/tags/tag_hr.asp) (`<em>`, `<strong>`, `<hr>`, `<sub>`, `<s>`)  
  

>   - Read about HTML [forms](https://www.w3schools.com/html/html_forms.asp)  

Reviewer: add forms to the `Instructions`

  
  </details>
  
<details>
     <summary>See the solution (only when done!)</summary>  
  
  https://codepen.io/kira-abl/pen/mddYrjj?editors=1000
  
</details>

### Task 4 - More HTML tags

**Tables**

In html a table is defined with the `<table>` tag. Each row is defined with the `<tr>` tag (**t**able **r**ow). A **t**able **h**eader is defined with the `<th>` tag. Each cell is defined with the `<td>` (**t**able **d**ata) tag.

> * Read more about [tables](https://www.w3schools.com/html/html_tables.asp), in a new .html file build a table that will have the following information in it:

Reviewer 2: ```suggestion
* Read more about [tables](https://www.w3schools.com/html/html_tables.asp), and in a new .html file build a table that will have the following information in it:
```

![](https://i.imgur.com/u9jlXqU.png)

> **HTML `<meta>` Tag**

Reviewer: this is an advanced topic, add it in the `Advanced Topics` section as a link.

Author: Do you mean a link in the section or a link in general?
I think the description should stay but under a heading of advanced

Reviewer: Yeah an Advanced heading would be great instead of being in `Task 4`. 
But then you dont need the description just have a title and the link. 
But if you want to keep the description then fine.

Author: Add the heading and keep the description :+1:

Reviewer 3: Okay, those are now under Advanced. The descriptions and instructions are there.

> Meta tags are snippets of text that describe a page’s content; they will not be visible on the page itself, but only in the page’s source code (html). Those tags are content descriptors that help tell search engines what a web page is about.Meta tags always go inside the `<head>` element.

Reviewer 2: ```suggestion
Meta tags are snippets of text that describe a page’s content; they will not be visible on the page itself, but only in the page’s source code (html). Those tags are content descriptors that help tell search engines what a web page is about. Meta tags always go inside the `<head>` element.
```

> * Read more about the [`<meta>`](https://www.w3schools.com/tags/tag_meta.asp) tag, in the file you are working on add a <meta> element that gives the browser instructions on how to control the page's dimensions and scaling.

Reviewer 2: ```suggestion
* Read more about the [`<meta>`](https://www.w3schools.com/tags/tag_meta.asp) tag, and in the file you are working on add a `<meta>` element that gives the browser instructions on how to control the page's dimensions and scaling.
```

<details>
     <summary>Hint</summary>  
Setting The Viewport
  </details>

> **HTML `<video>` Tag**  

Reviewer: this is an advanced topic, add it in the `Advanced Topics` section as a link.

Reviewer 3: Same here.
